packages/rudy/package.json

```diff
@@ -35,6 +35,8 @@
   },
   "homepage": "https://github.com/faceyspacey/redux-first-router#readme",
   "dependencies": {
+    "@respond-framework/middleware-change-page-title": "^0.1.0",
```

Comment:
yesssss, this is excellent, also out first installable middleware?

Can you quickly explain? How much access were you able to achieve with ur middleware. and how can we let developers externally tack on their own- something like this, but without access to our repo? 

Some areas here, you are surpassing me - which is really great to see and its the reason ur the only other maintainer on any faceyspacey project (other than me). As i was a godsend to James on Universal, you are a godsend to me on Rudy

Reply:
Haha, I do my best :)

For this one it was pretty simple - there was no tight coupling between the middleware and the core, so I basically just moved the code to a different directory. Absolutely, its a totally separate package - it just happens to be included in the default value of `middlewares` in `createRouter` (but sure, somebody could fork it or make something similar and pass their one into `createRouter`).

I also removed the `getTitle` from the core api passed to middlewares, because I figured getting the correct page title out of the redux state is a concern specific to this middleware, and it will make things easier to understand if the core is as small as possible. That's the only real change in structure.

I'd like to do the same to the other built in middlewares - this was just the easiest one. the `call()` middleware for example looks like its coupled to the core in a few ways so they will have to be decoupled first which will be a bit harder.

Reply:
Does call need to be decoupled. I always viewed it as a core part of just running can calling events. 

I’d leave true worse to last. As long as we get more access to the api, via middleware - that’s really great. And you have a example too. Some comments would go a long way or I’ll look at flow types also.

Reply:
Well none of it **needs** to be decoupled, but I think it would be good for style and maintainability if it was. Whether its a separate package or not is a detail, but I think if some functionality can be made into a middleware that touches the core only through a simple and well defined interface, then that's definitely better than having undocumented tight coupling between the two.

It would also help with customisation. To take one example, at the moment the `shouldCall` function is a parameter to the core `createRouter`, so its baked into the core API, can't be changed without a breaking change to the core, makes testing of the core that little bit harder, etc. But its only relevant to the `call` middleware.

I'm thinking instead, to add extra options `server`, `hydrate`, `browser` to the call middleware. Thta way the current default middlewares:

```javascript
[
    serverRedirect, // short-circuiting middleware
    anonymousThunk,
    pathlessRoute('thunk'),
    transformAction, // pipeline starts here
    call('beforeLeave', { prev: true }),
    call('beforeEnter'),
    enter,
    changePageTitle(),
    call('onLeave', { prev: true }),
    call('onEnter'),
    call('thunk', { cache: true }),
    call('onComplete'),
  ],
```

Would look instead like this:

```javascript
[
    serverRedirect, // short-circuiting middleware
    anonymousThunk,
    pathlessRoute('thunk'),
    transformAction, // pipeline starts here
    call('beforeLeave', { prev: true, server: false, hydrate: false }),
    call('beforeEnter', { hydrate: false }),
    enter,
    changePageTitle(),
    call('onLeave', { prev: true, server: false, hydrate: false }),
    call('onEnter', { server: false }),
    call('thunk', { cache: true, hydrate: false }),
    call('onComplete'),
  ],
```

That makes it easier to customise (you can change the names of the callbacks or decide which ones happen on the server or the client without patching the core), and I think also easier to read (the parameters make it obvious what the intention is re ssr and hydration, vs having to dig into the core code).

Another way of thinking about it: if our own included/built in middleware needs special treatment by the core, how is a third party developer going to build similar functionality? If we ensure that all the built in middlewares are decoupled and use the same API that is exposed to 3rd parties, then we're more likely to have a more complete/flexible API.

Reply:
I still need to get flow working lol - I tried a while back but ran into a sea of problems with vscode, and unfortunately I really like vscode otherwise...

Reply:
and yeah, "worst till last" sounds ideal. Basically I'm thinking if I can chip away at small improvements/simplifications like this that I can easily understand, a bigger task like nested routes will be easier, because there will be less tight coupled things to worry about, and I will have learned more about the core code in the process of doing the easier things.